src/compiler/utilities.ts

```diff
@@ -6258,6 +6258,29 @@ namespace ts {
         return isSourceFile(node) || isModuleBlock(node) || isBlock(node) && isFunctionLike(node.parent);
     }
 
+    /* @internal */
+    export function isNewThisScope(node: Node): boolean {
+        switch (node.kind) {
+            case SyntaxKind.ClassDeclaration:
+            case SyntaxKind.ClassExpression:

```

Comment:
`class C extends this.OtherClass {}` the `this` actually refers to the containing object or class.
The same applies to computed names of methods and accessors in that class (although TypeScript currently doesn't allow that)